src/main/flight/gps_rescue.c

```diff
@@ -0,0 +1,373 @@
+    rescueState.sensor.maxAltitude = MAX(rescueState.sensor.currentAltitude, rescueState.sensor.maxAltitude);
+    // Store the max distance to home during normal flight so we know if a flyaway is happening
+    rescueState.sensor.maxDistanceToHome = MAX(rescueState.sensor.distanceToHome, rescueState.sensor.maxDistanceToHome);
+
+    rescueThrottle = rcCommand[THROTTLE];
+
+    //to do: have a default value for hoverThrottle
+    float ct = getCosTiltAngle();
+    if (ct > 0.5 && ct < 0.96 && throttleSamples < 1E6 && rescueThrottle > 1070) { //5 to 45 degrees tilt
+        //TO DO: only sample when acceleration is low
+        uint16_t adjustedThrottle = 1000 + (rescueThrottle - 1000) * ct;
+        if (throttleSamples == 0) {
+            averageThrottle = adjustedThrottle;
+        } else {
+            averageThrottle += (adjustedThrottle - averageThrottle) / (throttleSamples + 1);
+        }
+        hoverThrottle = lrintf(averageThrottle);
+        throttleSamples++;
+    }
+
+}
+
+void rescueAttainPosition()
+{
+    // Point to home if that is in our intent
+    if (rescueState.intent.crosstrack) {
+        setBearing(rescueState.sensor.directionToHome);
+    }
+
+    if (!newGPSData) {
+        return;
+    }
+
+    /**
+        Speed controller
+    */
+    static float previousSpeedError = 0;
+    static int16_t speedIntegral = 0;
+
+    const int16_t speedError = (rescueState.intent.targetGroundspeed - rescueState.sensor.groundSpeed) / 100;
+    const int16_t speedDerivative = speedError - previousSpeedError;
+
+    speedIntegral = constrain(speedIntegral + speedError, -100, 100);
+
+    previousSpeedError = speedError;
+
+    int16_t angleAdjustment =  gpsRescue()->vP * speedError + (gpsRescue()->vI * speedIntegral) / 100 +  gpsRescue()->vD * speedDerivative;
+
+    gpsRescueAngle[AI_PITCH] = constrain(gpsRescueAngle[AI_PITCH] + MIN(angleAdjustment, 80), rescueState.intent.minAngle * 100, rescueState.intent.maxAngle * 100);
+
+    float ct = cos(DECIDEGREES_TO_RADIANS(gpsRescueAngle[AI_PITCH] / 10));
+
+    /**
+        Altitude controller
+    */
+    static float previousAltitudeError = 0;
+    static int16_t altitudeIntegral = 0;
+
+    const int16_t altitudeError = (rescueState.intent.targetAltitude - rescueState.sensor.currentAltitude) / 100; // Error in meters
+    const int16_t altitudeDerivative = altitudeError - previousAltitudeError;
+
+    // Only allow integral windup within +-15m absolute altitude error
+    if (ABS(altitudeError) < 25) {
+        altitudeIntegral = constrain(altitudeIntegral + altitudeError, -250, 250);
+    } else {
+        altitudeIntegral = 0;
+    }
+
+    previousAltitudeError = altitudeError;
+
+    int16_t altitudeAdjustment = (gpsRescue()->tP * altitudeError + (gpsRescue()->tI * altitudeIntegral) / 10 *  + gpsRescue()->tD * altitudeDerivative) / ct / 20;
+    int16_t hoverAdjustment = (hoverThrottle - 1000) / ct;
```

Comment:
1000 is a magic number. It's better to use a meaningful literal `PWM_RANGE_MIN`

Reply:
Will fix this 👍

Reply:
Fixed.